bokehjs/src/lib/models/annotations/info_pane.ts

```diff
@@ -0,0 +1,207 @@
+
+    display(this.el)
+
+   // slightly confusing: side "left" (for example) is relative to point that
+   // is being annotated but CS class ".bk-left" is relative to the tooltip itself
+
+    let top: number = 0
+    let left = 0
+    let right = 0
+
+    switch (anchor) {
+      case "left":
+        this.el.classList.add(bk_right)
+        left = x + (this.el.offsetWidth - this.el.clientWidth)
+        top = y - this.el.offsetHeight/2
+        break
+      case "right":
+        this.el.classList.add(bk_left)
+        right = (this.plot_view.layout.bbox.width - x)
+        top = y - this.el.offsetHeight/2
+        break
+      case "above":
+        this.el.classList.add(bk_below)
+        top = y + (this.el.offsetHeight - this.el.clientHeight) - arrow_size
+        left = Math.round(x - this.el.offsetWidth/2)
+        break
+      case "below":
+        this.el.classList.add(bk_above)
+        top = y - this.el.offsetHeight + arrow_size
+        left = Math.round(x - this.el.offsetWidth/2)
+        break
+    }
+
+    if (this.model.show_arrow)
+      this.el.classList.add(bk_tooltip_arrow)
+
+    if (this.el.childNodes.length > 0) {
+      this.el.style.top = `${top}px`
+      this.el.style.left = left ? `${left}px` : 'auto'
+      this.el.style.right = right ? `${right}px` : 'auto'
+    } else
+      undisplay(this.el)
+  }
+}
+
+export namespace InfoPane {
+  export type Attrs = p.AttrsOf<Props>
+
+  export type Props = Annotation.Props & {
+    anchor: p.Property<TooltipAttachment>
+    //can be ignored if inner_only property isnt useful
+    inner_only: p.Property<boolean>
+    show_arrow: p.Property<boolean>
+    x: p.Property<number>
+    y: p.Property<number>
+    anchor_units: p.Property<SpatialUnits>
+    //data has to be able to get list of strings
+    data: p.Property<[string][]>
+    custom: p.Property<boolean>
+  }
+}
+
+export interface InfoPane extends InfoPane.Attrs {}
+
+export class InfoPane extends Annotation {
+  properties: InfoPane.Props
+  __view_type__: InfoPaneView
+
+  constructor(attrs?: Partial<InfoPane.Attrs>) {
+    super(attrs)
+  }
+
+  static init_InfoPane(): void {
+    this.prototype.default_view = InfoPaneView
+
+    this.define<InfoPane.Props>({
+      //is kept as is if the horizontal\vertical functionality shall come in use
+      anchor:       [ p.TooltipAttachment, 'horizontal'],
+      inner_only:   [ p.Boolean,           true        ],
+      show_arrow:   [ p.Boolean,           true        ],
+      x:            [ p.Number,             0          ],
+      y:            [ p.Number,             0          ],
+      anchor_units: [ p.SpatialUnits,      'data'      ],
+      data:         [ p.Array,                         ],
+    })
+
+    this.override({
+      level: 'overlay',
+    })
+
+    this.internal({
+      custom: [ p.Any     ],
+    })
+  }
+
+  clear(): void {
+    this.data = []
+  }
+
+  add(pos_x: number, pos_y: number, content: string): void {
```

Comment:
So I don't have any objection to making `.data` an array, I do find this API a little odd though. Setting the position every time you add an extra tooltip section seems awkward. What's this method used by?

Reply:
This is the way the hover tool currently sets its internal `ToolTip` I would suggest we jettison this and re-visit any specializations or API needed for HoverTool usage anew in the future.